src/tower_actor/actor.rs

```diff
@@ -0,0 +1,96 @@
+use std::{
+    future::Future,
+    task::{Context, Poll},
+};
+
+use crate::tower_actor::future::ActorFuture;
+use crate::tower_actor::message::{Message, Request};
+
+use thiserror::Error;
+use tokio::sync::{mpsc, oneshot};
+use tokio_util::sync::PollSender;
+use tower::Service;
+
+#[derive(Error, Debug)]
+pub enum ActorError {
+    #[error("Actor has terminated or panicked")]
+    ActorTerminated,
+    #[error("Actor has dropped this request without responding")]
+    ActorHungUp,
+}
+
+pub struct Actor<R, S, E> {
+    queue: tokio_util::sync::PollSender<Message<R, S, E>>,
+}
+
+impl<R, S, E> Actor<R, S, E>
+where
+    R: Request + 'static,
+    S: Send + 'static,
+    E: Send + 'static,
+{
+    pub fn new<F, W>(bound: usize, f: F) -> Self
+    where
+        F: FnOnce(mpsc::Receiver<Message<R, S, E>>) -> W,
+        W: Future<Output = Result<(), E>> + Send + 'static,
+    {
+        Self::named("tower-actor-worker", bound, f)
+    }
+
+    pub fn named<'a, F, W>(name: &'a str, bound: usize, f: F) -> Self
+    where
+        F: FnOnce(mpsc::Receiver<Message<R, S, E>>) -> W,
+        W: Future<Output = Result<(), E>> + Send + 'static,
+    {
+        let (queue_tx, queue_rx) = mpsc::channel(bound);
+
+        tokio::task::Builder::new()
+            .name(name)
+            .spawn(f(queue_rx))
+            .expect("failed to spawn worker");
+
+        Self {
+            queue: PollSender::new(queue_tx),
+        }
+    }
+}
+
+impl<R, S, E> Service<R> for Actor<R, S, E>
+where
+    R: Request + 'static,
+    S: Send + 'static,
+    E: Send + 'static + From<ActorError>,
+{
+    type Response = S;
+    type Error = E;
+    type Future = ActorFuture<S, E>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.queue
+            .poll_reserve(cx)
+            .map_err(|_e| ActorError::ActorTerminated.into())
+    }
+
+    fn call(&mut self, req: R) -> Self::Future {
+        // Due to the permit system that PollSender uses, we can always send on the queue if
+        // Poll reserve succeeded.
+        // See: https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.close
+        //
+        // Since the Service contract requires that `poll_ready()` pass
+        // before calling `call()`, we can safely proceed without checking that the queue isn't closed.
+        debug_assert!(!self.queue.is_closed());
+
+        let span = req.create_span();
```

Comment:
I think this is a holdover from the pd code, which hooked custom span creation into the service definitions. But in a generic context I don't think it's appropriate for the service to create custom spans as well as doing the actor part — we should just use Span::current() here, and if the caller wants to do custom spam handling, they should do that by composing in an additional tower layer. That way, we only do one thing, and let the user compose behavior as desired.